File: backend/src/controllers/authController.js
import { pool } from '../config/database.js';

// Mock de login (sin autenticación real por ahora)
export const login = async (req, res) => {
  try {
    const { correo, contrasena } = req.body;

    if (!correo || !contrasena) {
      return res.status(400).json({
        error: 'Correo y contraseña son requeridos'
      });
    }

    // Buscar el login
    const [loginRows] = await pool.query(
      'SELECT * FROM login WHERE correo = ?',
      [correo]
    );

    if (loginRows.length === 0) {
      return res.status(401).json({
        error: 'Credenciales inválidas'
      });
    }

    const login = loginRows[0];

    // En producción aquí se compararía con bcrypt
    // Por ahora comparación directa (MOCK)
    if (login.contrasena !== contrasena) {
      return res.status(401).json({
        error: 'Credenciales inválidas'
      });
    }

    // Buscar datos del participante usando el email del login
    const [participanteRows] = await pool.query(
      'SELECT ci, nombre, apellido, email FROM participante WHERE email = ?',
      [correo]
    );

    if (participanteRows.length === 0) {
      return res.status(404).json({
        error: 'Participante no encontrado'
      });
    }

    const participante = participanteRows[0];
    
    // Obtener el rol del participante desde participante_programa_académico
    const [rolRows] = await pool.query(
      'SELECT rol FROM participante_programa_académico WHERE ci = ? LIMIT 1',
      [participante.ci]
    );
    
    const rol = rolRows.length > 0 ? rolRows[0].rol : null;

    // Mock token (en producción sería JWT)
    const token = `mock-token-${participante.ci}-${Date.now()}`;

    res.json({
      success: true,
      token,
      participante: {
        ci: participante.ci,
        nombre: participante.nombre,
        apellido: participante.apellido,
        email: participante.email,
        rol: rol
      }
    });

  } catch (error) {
    console.error('Error en login:', error);
    console.error('Error stack:', error.stack);
    res.status(500).json({
      error: 'Error en el servidor',
      details: process.env.NODE_ENV === 'development' ? error.message : undefined
    });
  }
};

// Mock logout
export const logout = async (req, res) => {
  try {
    // En producción aquí se invalidaría el token
    res.json({
      success: true,
      message: 'Sesión cerrada exitosamente'
    });
  } catch (error) {
    console.error('Error en logout:', error);
    res.status(500).json({
      error: 'Error en el servidor'
    });
  }
};

// Mock me (obtener usuario actual)
export const me = async (req, res) => {
  try {
    // Por ahora mockeado, obtener CI desde query o header
    const ci = req.query.ci || req.headers['x-user-ci'];

    if (!ci) {
      return res.status(401).json({
        error: 'No autenticado'
      });
    }

    const [participanteRows] = await pool.query(
      'SELECT ci, nombre, apellido, email FROM participante WHERE ci = ?',
      [ci]
    );

    if (participanteRows.length === 0) {
      return res.status(404).json({
        error: 'Participante no encontrado'
      });
    }

    const participante = participanteRows[0];
    
    // Obtener el rol del participante desde participante_programa_académico
    const [rolRows] = await pool.query(
      'SELECT rol FROM participante_programa_académico WHERE ci = ? LIMIT 1',
      [ci]
    );
    
    const rol = rolRows.length > 0 ? rolRows[0].rol : null;

    res.json({
      success: true,
      participante: {
        ci: participante.ci,
        nombre: participante.nombre,
        apellido: participante.apellido,
        email: participante.email,
        rol: rol
      }
    });

  } catch (error) {
    console.error('Error en me:', error);
    console.error('Error stack:', error.stack);
    res.status(500).json({
      error: 'Error en el servidor',
      details: process.env.NODE_ENV === 'development' ? error.message : undefined
    });
  }
};

File: backendPython/src/controllers/auth_controller.py
from fastapi import HTTPException
from src.config.database import get_connection
from datetime import datetime

async def login(correo: str, contrasena: str):
    """Autenticación de usuario"""
    try:
        if not correo or not contrasena:
            raise HTTPException(
                status_code=400,
                detail="Correo y contraseña son requeridos"
            )

        conn = get_connection()
        cursor = conn.cursor(dictionary=True)

        # Buscar el login
        cursor.execute('SELECT * FROM login WHERE correo = %s', (correo,))
        login_rows = cursor.fetchall()

        if len(login_rows) == 0:
            cursor.close()
            conn.close()
            raise HTTPException(
                status_code=401,
                detail="Credenciales inválidas"
            )

        login_data = login_rows[0]

        # En producción aquí se compararía con bcrypt
        # Por ahora comparación directa (MOCK)
        if login_data['contrasena'] != contrasena:
            cursor.close()
            conn.close()
            raise HTTPException(
                status_code=401,
                detail="Credenciales inválidas"
            )

        # Buscar datos del participante usando el email del login
        cursor.execute(
            'SELECT ci, nombre, apellido, email FROM usuario WHERE email = %s',
            (correo,)
        )
        participante_rows = cursor.fetchall()

        if len(participante_rows) == 0:
            cursor.close()
            conn.close()
            raise HTTPException(
                status_code=404,
                detail="Participante no encontrado"
            )

        participante = participante_rows[0]

        # Obtener el rol del participante desde participante_carrera
        cursor.execute(
            'SELECT rol FROM participante_carrera WHERE ci = %s LIMIT 1',
            (participante['ci'],)
        )
        rol_rows = cursor.fetchall()

        rol = rol_rows[0]['rol'] if len(rol_rows) > 0 else None

        cursor.close()
        conn.close()

        # Mock token (en producción sería JWT)
        token = f"mock-token-{participante['ci']}-{int(datetime.now().timestamp() * 1000)}"

        return {
            "success": True,
            "token": token,
            "participante": {
                "ci": participante['ci'],
                "nombre": participante['nombre'],
                "apellido": participante['apellido'],
                "email": participante['email'],
                "rol": rol
            }
        }

    except HTTPException:
        raise
    except Exception as error:
        print(f'Error en login: {error}')
        raise HTTPException(
            status_code=500,
            detail="Error en el servidor"
        )


async def logout():
    """Cerrar sesión"""
    try:
        # En producción aquí se invalidaría el token
        return {
            "success": True,
            "message": "Sesión cerrada exitosamente"
        }
    except Exception as error:
        print(f'Error en logout: {error}')
        raise HTTPException(
            status_code=500,
            detail="Error en el servidor"
        )


async def me(ci: str = None):
    """Obtener información del usuario actual"""
    try:
        if not ci:
            raise HTTPException(
                status_code=401,
                detail="No autenticado"
            )

        conn = get_connection()
        cursor = conn.cursor(dictionary=True)

        cursor.execute(
            'SELECT ci, nombre, apellido, email FROM usuario WHERE ci = %s',
            (ci,)
        )
        participante_rows = cursor.fetchall()

        if len(participante_rows) == 0:
            cursor.close()
            conn.close()
            raise HTTPException(
                status_code=404,
                detail="Participante no encontrado"
            )

        participante = participante_rows[0]

        # Obtener el rol del participante desde participante_carrera
        cursor.execute(
            'SELECT rol FROM participante_carrera WHERE ci = %s LIMIT 1',
            (ci,)
        )
        rol_rows = cursor.fetchall()

        rol = rol_rows[0]['rol'] if len(rol_rows) > 0 else None

        cursor.close()
        conn.close()

        return {
            "success": True,
            "participante": {
                "ci": participante['ci'],
                "nombre": participante['nombre'],
                "apellido": participante['apellido'],
                "email": participante['email'],
                "rol": rol
            }
        }

    except HTTPException:
        raise
    except Exception as error:
        print(f'Error en me: {error}')
        raise HTTPException(
            status_code=500,
            detail="Error en el servidor"
        )

